crates/schema/src/def.rs

```diff
@@ -0,0 +1,640 @@
+//!
+//! After validation, a `ModuleDef` can be converted to the `*Schema` types in `crate::schema` for use in the database.
+//! (Eventually, we may unify these types...)
+
+use std::fmt::{self, Write};
+
+use crate::error::{IdentifierError, ValidationErrors};
+use crate::identifier::Identifier;
+use itertools::Itertools;
+use spacetimedb_data_structures::error_stream::{CollectAllErrors, CombineErrors, ErrorStream};
+use spacetimedb_data_structures::map::HashMap;
+use spacetimedb_lib::db::auth::{StAccess, StTableType};
+use spacetimedb_lib::db::raw_def;
+use spacetimedb_lib::db::raw_def::v9::{RawIdentifier, RawScopedTypeNameV9};
+use spacetimedb_lib::{ProductType, RawModuleDef};
+use spacetimedb_primitives::{ColId, ColList};
+use spacetimedb_sats::AlgebraicType;
+use spacetimedb_sats::{AlgebraicTypeRef, Typespace};
+
+pub mod validate;
+
+/// A map from `Identifier`s to values of type `T`.
+pub type IdentifierMap<T> = HashMap<Identifier, T>;
+
+// We may eventually want to reorganize this module to look more
+// like the system tables, with numeric IDs used for lookups
+// in addition to `Identifier`s.
+//
+// If that path is taken, it might be possible to have this type
+// entirely subsume the various `Schema` types, which would be cool.
+
+/// A validated, canonicalized, immutable module definition.
+///
+/// Cannot be created directly. Instead, create/deserialize a [spacetimedb_lib::RawModuleDef] and call [ModuleDef::validate]
+/// or use `try_into`.
+///
+/// ```rust
+/// use spacetimedb_lib::RawModuleDef;
+/// use spacetimedb_schema::def::{ModuleDef, TableDef, IndexDef, TypeDef, ModuleDefLookup, ScopedTypeName};
+/// use spacetimedb_schema::identifier::Identifier;
+///
+/// fn read_raw_module_def_from_file() -> RawModuleDef {
+///     // ...
+/// #   RawModuleDef::V9(Default::default())
+/// }
+///
+/// let raw_module_def = read_raw_module_def_from_file();
+/// let module_def = ModuleDef::validate(raw_module_def).expect("valid module def");
+///
+/// let table_name = Identifier::new("my_table".into()).expect("valid identifier");
+/// let index_name = Identifier::new("my_index".into()).expect("valid identifier");
+/// let scoped_type_name = ScopedTypeName::try_new([], "MyType").expect("valid scoped type name");
+///
+/// let table: Option<&TableDef> = module_def.lookup(&table_name);
+/// let index: Option<&IndexDef> = module_def.lookup(&index_name);
+/// let type_def: Option<&TypeDef> = module_def.lookup(&scoped_type_name);
+/// // etc.
+/// ```
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct ModuleDef {
+    /// The tables of the module definition.
+    tables: IdentifierMap<TableDef>,
+
+    /// The reducers of the module definition.
+    reducers: IdentifierMap<ReducerDef>,
+
+    /// The type definitions of the module definition.
+    types: HashMap<ScopedTypeName, TypeDef>,
+
+    /// The typespace of the module definition.
+    typespace: Typespace,
```

Comment:
When we add `AlgebraicTypeUse` and `AlgebraicTypeDefinition`, we may want to change this to a datastructure that stores those for client codegen. On the other hand, I think the database engine still wants to think about algebraic types structurally, so we may be in the uncomfortable position of needing both.

Reply:
I think in the future we will want to think about how we can implement nominal types within a structural type system on a whollistic level.

Reply:
Yeah, I think that would be very worthwhile.